test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
   testWithSize(37, 11);
 }
 
+static std::unordered_map<const Buf*, TensorAccessBoundsInfo>
+convertBoundsInfoToMap(const std::vector<TensorAccessBoundsInfo>& v) {
+  std::unordered_map<const Buf*, TensorAccessBoundsInfo> res;
+  for (const auto& el : v) {
+    res[el.buf] = el;
+  }
+  return res;
+}
+
+static void verifyConstBounds(
+    const TensorAccessBoundsInfo& access_info,
+    const std::vector<std::pair<int, int>>& ref) {
+  size_t ndim = ref.size();
+  ASSERT_EQ(access_info.start.size(), ndim);
+  ASSERT_EQ(access_info.stop.size(), ndim);
+  for (size_t i = 0; i < ndim; i++) {
+    if (ref[i].first >= 0) { // Negative values are used to skip the check
+      auto start_imm = dynamic_cast<const IntImm*>(access_info.start[i]);
+      ASSERT_TRUE(start_imm);
+      ASSERT_EQ(start_imm->value(), ref[i].first);
+    }
+    if (ref[i].second >= 0) {
+      auto stop_imm = dynamic_cast<const IntImm*>(access_info.stop[i]);
+      ASSERT_TRUE(stop_imm);
+      ASSERT_EQ(stop_imm->value(), ref[i].second);
+    }
+  }
+}
+
+void testBoundsInference_1() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..100:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 99}}
+  KernelScope kernel_scope;
+  ExprHandle n(100);
+  Buffer a(BufHandle("a", {n}), kFloat);
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
+      inferBounds(l.root_stmt());
+  auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, 99}});
+  ASSERT_EQ(bounds_info_map.at(b->buf()).kind, kStore);
+  verifyConstBounds(bounds_info_map.at(b->buf()), {{0, 99}});
+}
+
+void testBoundsInference_2() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..n:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, n-1}, {a, kLoad, 0, n-1}}
+  KernelScope kernel_scope;
+  VarHandle n("n", kInt);
+  Buffer a(BufHandle("a", {n}), kFloat);
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
+      inferBounds(l.root_stmt());
+  auto bounds_info_map = convertBoundsInfoToMap(bounds_info);
+
+  // We should have two entries: one for 'b' and one for 'a'.
+  ASSERT_EQ(bounds_info_map.size(), 2);
+  ASSERT_EQ(bounds_info_map.at(a.data()).kind, kLoad);
+  verifyConstBounds(bounds_info_map.at(a.data()), {{0, -1}});
```

Comment:
"-1" looks a bit weird to imply "None". Most system use "None", which is a particular very negative number. Or an OptionalNumber where it can be implicitly constructed from an integer, or None which implies that there is no number.

Reply:
This is only for these three tests, I would prefer not to over-engineer this.